lib/authentication/sso_url_provider.js

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2015-2023 Snowflake Computing Inc. All rights reserved.
+ */
+
+const Util = require('../util');
+const Errors = require('../errors');
+const { rest } = require('../global_config');
+
+/**
+ * Creates a new instance of an SsoUrlProvider.
+ *
+ * @param {Object} connectionConfig
+ * @param {Object} httpClient
+ * @constructor
+ */
+function SsoUrlProvider(connectionConfig, httpClient) {
+
+  Errors.assertInternal(Util.isObject(connectionConfig));
+  Errors.assertInternal(Util.isObject(httpClient));
+
+  const port = rest.HTTPS_PORT;
+  const protocol = rest.HTTPS_PROTOCOL;
+
+  /**
+   * Get SSO URL through POST request.
+   *
+   * @param {String} authenticator
+   * @param {String} serviceName
+   * @param {String} account
+   * @param {Number} callbackPort
+   * @param {String} user
+   * @param {String} host
+   *
+   * @returns {String} the SSO URL.
+   */
+  this.getSSOURL = function (authenticator, serviceName, account, callbackPort, user, host) {
+    // Create URL to send POST request to
+    const url = protocol + '://' + host + '/session/authenticator-request';
+
+    let header;
+    if (serviceName) {
+      header = {
+        'HTTP_HEADER_SERVICE_NAME': serviceName
+      };
+    }
+    const body = {
+      'data': {
+        'ACCOUNT_NAME': account,
+        'LOGIN_NAME': user,
+        'PORT': port,
+        'PROTOCOL': protocol,
+        'AUTHENTICATOR': authenticator,
+        'BROWSER_MODE_REDIRECT_PORT': callbackPort.toString()
+      }
+    };
+
+    const agent = httpClient.getAgent(url, connectionConfig.getProxy());
+
+    const requestOptions =
+      {
+        headers: header,
+        requestOCSP: false,
+        rejectUnauthorized: true,
+        httpsAgent: agent
```

Comment:
do we need explicitly setting agent here? our http client should already be proxy aware?

Reply:
but we have custom method to create agent `getAgent`. It contains all conditions like ocsp, proxy ,mock

Reply:
It is set inside httpClient now